Adding MWP conversion changes.

michaelferrara opened the pull request:
Made multiple changes to convert the my widgets page from AngularJS to React. Worked on the support page and header a little as well.

## FUTURE TODOS? 
A list of stuff to not forget about, if they aren't in the scope of this PR we should make issues out of them...
* [ ] separate support and admin menus, redesign to fit better with the existing look and feel (but still draw attention)
* [ ] set up Jest and or Cypress and get some tests in place
* [ ] code quality tools, prettier, eslint, csslint, husky, auto-run tests etc.
* [ ] Optimize cached responses using browser localStorage.

Commit:
Add some line to inform user about canvas integration with link

clpetersonucf committed:
Initial implementation

clpetersonucf committed:
Additions, added confirmation dialog

clpetersonucf committed:
Still WIP, confirmation dialog to be reworked

clpetersonucf committed:
Removed rollback confirm dialog, adding action bar modification to support its role

clpetersonucf committed:
Refinements

clpetersonucf committed:
Added test for new API endpoint

clpetersonucf committed:
Merge branch 'dev/7-alexander' into issue/1223-view-previous-saves
Merging in dev/7 updates

clpetersonucf committed:
Removed debug trace statement

iturgeon committed:
implements dotenv config changes

iturgeon committed:
minor polish on env vars

iturgeon committed:
dotenv iteration - environments working

iturgeon committed:
adds .env.pristine

iturgeon committed:
rolls crypt.php into new env config

iturgeon committed:
moved to .env + .env.local config model

iturgeon committed:
adds test env and simplifies use of FUEL_ENV

iturgeon committed:
add getting started to default help page

iturgeon committed:
lti config supports default, more env config

* lti config now supports a 'default' consumer that will fall back on if match not found
* settled on symfony dotenv
* .env.local used for docker dev and docker test envs
* log handler resolver upgraded in both places
* lti test provider link added to /dev

iturgeon committed:
fix codesnif rules and resulting issues

iturgeon committed:
fix fuelphp-phpcs to a tagged version

iturgeon committed:
remove config wizard

iturgeon committed:
minified .env and moved description to readme

iturgeon committed:
[update] lti embed preview page updates

* preview pre-page now shows widget name and widget type icon
* preview pre-page opens widget in embedded mode now
* preview supports embed mode
* NOTE requires updates to MSCA (issue/preview-with-embed branch)

iturgeon committed:
spelling mistake in README.md

Co-authored-by: Brandon Stull <bstull@ucf.edu>

iturgeon committed:
spelling fix in README.md

Co-authored-by: Brandon Stull <bstull@ucf.edu>

dependabot[bot] committed:
Bump decompress from 4.2.0 to 4.2.1

Bumps [decompress](https://github.com/kevva/decompress) from 4.2.0 to 4.2.1.
- [Release notes](https://github.com/kevva/decompress/releases)
- [Commits](https://github.com/kevva/decompress/compare/v4.2.0...v4.2.1)

Signed-off-by: dependabot[bot] <support@github.com>

iturgeon committed:
Docker readme updates

iturgeon committed:
Merge remote-tracking branch 'upstream/dev/7.1' into issue/1112-streamline-install-and-config

iturgeon committed:
adds flag to prevent lti default config and updates docker https for assets

iturgeon committed:
Merge branch 'issue/preview-with-embed' into issue/1112-streamline-install-and-config

iturgeon committed:
moves php config vars from nginx to php.ini

iturgeon committed:
adds optional lti logging

iturgeon committed:
finalizes support for lti launch single score screens

iturgeon committed:
code style updates

iturgeon committed:
removes trace from add_log

iturgeon committed:
clean test output

ljoks committed:
creates support page and its search component

ljoks committed:
adds ability to search for widget instances

ljoks committed:
add is_deleted attribute to widget instance class

ljoks committed:
accounts for no content response in instance search

ljoks committed:
adds component for selected instance in support dashboard, also adds 'is_deleted' attribute to widget instance class

ljoks committed:
adds functionality to 'undelete' a widget instance

ljoks committed:
fixes undelete and implements copying instances

ljoks committed:
remove console logs

clpetersonucf committed:
Merge pull request #1310 from iturgeon/docker-readme-updates

Docker readme updates

clpetersonucf committed:
Merge pull request #1308 from ucfopen/dependabot/npm_and_yarn/decompress-4.2.1

Bump decompress from 4.2.0 to 4.2.1

ljoks committed:
stable date handling and instance updating

ljoks committed:
creates support role

ljoks committed:
include 'Context' for instances played

ljoks committed:
Merge remote-tracking branch 'ian/issue/combine-msca-and-convert-to-react' into issue/support-dashboard-react

iturgeon committed:
fixes modals and temp fix for react css showing up in wrong directory

clpetersonucf committed:
Changed reference to Canvas to generally reference your LMS

clpetersonucf committed:
Resolving merge conflict with dev/7.1

clpetersonucf committed:
Merge pull request #1260 from Goff-Davis/issue/1216-inline-help-my-widgets

#1216 Inline Help on My Widgets Page

clpetersonucf committed:
Merge branch 'dev/7.1' into issue/1112-streamline-install-and-config

ljoks committed:
updates to make the collab dialog work (on it's own and with support page)

clpetersonucf committed:
Adds requires lti_message_type param to allow tests to pass

clpetersonucf committed:
Merge pull request #1300 from iturgeon/issue/1112-streamline-install-and-config

Implements dotenv to streamline config  & install

clpetersonucf committed:
Merge branch 'dev/7.1' into issue/1223-view-previous-saves

clpetersonucf committed:
small changes based on feedback

clpetersonucf committed:
updated question_history_get test

clpetersonucf committed:
renamed classes and ids with underscores in qset history

clpetersonucf committed:
Replaced question_history_get in old v1 api with instance api controller call

clpetersonucf committed:
Replaced Previously Saved with Save History

ljoks committed:
finishes implementation for widget collab dialog in react

clpetersonucf committed:
Replaced playable_by_current_user with more specific user_has_any_perm_to check for history api call

clpetersonucf committed:
Replaced Api_V1 test with controller instance test

ljoks committed:
disables collab button if instance is deleted

clpetersonucf committed:
Merge pull request #1268 from clpetersonucf/issue/1223-view-previous-saves

View and restore previously saved qsets in creator

ljoks committed:
starts gui implementation, adds search

ljoks committed:
minor fixes

ljoks committed:
Merge branch 'issue/support-dashboard-react' of github.com:ljoks/Materia into extra-attempts-gui

clpetersonucf committed:
Bumped MSCA to 2.3.0-alpha1 for QA testing

ljoks committed:
adds visuals for component, just need to implement button functionality

ljoks committed:
adds functionality for getting and setting extra attempts

ljoks committed:
implements a working solution for extra attempts gui

ljoks committed:
adds comments and removes logs

ljoks committed:
adds comments, removes console logs

ljoks committed:
implements certain permissions for the support user role

ljoks committed:
linter errors

ljoks committed:
Merge pull request #1 from ljoks/extra-attempts-gui

Extra attempts gui

ljoks committed:
adds migration to create support user role in the db

ljoks committed:
Merge pull request #2 from ljoks/support-role-perms

Support role perms

ljoks committed:
adds Success! indicator if changes are successfully applied

michaelferrara committed:
My widgets will now load if the user has no widgets.

iturgeon committed:
adds GOOGLE_ANALYTICS_ID env var, docs for lti consumers

iturgeon committed:
enable env to control theme

michaelferrara committed:
Finished collaboration modal and made a few changes to the default modal css.

michaelferrara committed:
Added edit button functionality and finished upper half of controls section of the my widgets page.

michaelferrara committed:
Fixed perms for showing additional content on the selected widget instance.

michaelferrara committed:
Added warning modal when edit button is clicked on a selected widget instance, and added to perms on the same page

clpetersonucf committed:
Updates post_login docs links

clpetersonucf committed:
Replaced THEME_PATH with THEME_PACKAGE in docker env configs

iturgeon committed:
switch from travis to actions

iturgeon committed:
correct github actions name for tests

iturgeon committed:
add skip_check to tests to reduce duplicate action runs

iturgeon committed:
production ready docker changes

simplify docker updates by reducing to single docker container to maintain
use production ready docker container in dev
move env config vars entirely into docker-compose instead of .env
introduce multi-stage builds for production ready docker server
updates readme to add a few more hints and tips for configuration

iturgeon committed:
rn docker containers, nginx now produciton ready

* renames phpfpm docker container to app
* renames nginx docker container to webserver
* adds dockerfile for webserver
* sets docker-compose project name 'docker' prefex is now materia

iturgeon committed:
add yarn-error to gitignore

iturgeon committed:
fix dockerfile to be able to build from clean checkout

iturgeon committed:
runs app container as non-root user

iturgeon committed:
root-less docker containers

iturgeon committed:
fix syntax error in docker file multiline

iturgeon committed:
add more files to app dockerignore to speed up build

iturgeon committed:
move s3 docker file, some fixes for run_test_ci

iturgeon committed:
fix fakes3 docker file in docker compose

iturgeon committed:
fix docker test

adds a script to set uid on moounted files from the host for tests
adds a fuel/vendor/.gitkeep

iturgeon committed:
switch to module php.ini and enable xdebug for coverage

iturgeon committed:
enables xdebug.mode=coverage in test

FrenjaminBanklin committed:
Hotfix to no_attempt function definition/call - type expectations were incorrect for  and  was not being provided.

clpetersonucf committed:
Merge pull request #1316 from FrenjaminBanklin/issue/hotfix-controller-widgets-no-attempts

Hotfix no_attempt in widgets controller

clpetersonucf committed:
Spelling and grammatical corrections

Courtesy of @FrenjaminBanklin

Co-authored-by: Brandon Stull <bstull@ucf.edu>

clpetersonucf committed:
MSCA 2.3.0 and removed extranneous whitespace

clpetersonucf committed:
Updated THEME_PATH to THEME_PACKAGE in readme

clpetersonucf committed:
Merge pull request #1309 from ucfopen/dev/7.1

Dev/8.0

dependabot[bot] committed:
Bump ini from 1.3.5 to 1.3.8

Bumps [ini](https://github.com/isaacs/ini) from 1.3.5 to 1.3.8.
- [Release notes](https://github.com/isaacs/ini/releases)
- [Commits](https://github.com/isaacs/ini/compare/v1.3.5...v1.3.8)

Signed-off-by: dependabot[bot] <support@github.com>

clpetersonucf committed:
Merge pull request #1322 from ucfopen/dependabot/npm_and_yarn/ini-1.3.8

Bump ini from 1.3.5 to 1.3.8

dependabot[bot] committed:
Bump dot-prop from 4.2.0 to 4.2.1

Bumps [dot-prop](https://github.com/sindresorhus/dot-prop) from 4.2.0 to 4.2.1.
- [Release notes](https://github.com/sindresorhus/dot-prop/releases)
- [Commits](https://github.com/sindresorhus/dot-prop/compare/v4.2.0...v4.2.1)

Signed-off-by: dependabot[bot] <support@github.com>

clpetersonucf committed:
Merge pull request #1323 from ucfopen/dependabot/npm_and_yarn/dot-prop-4.2.1

Bump dot-prop from 4.2.0 to 4.2.1

clpetersonucf committed:
v8.0.0

iturgeon committed:
adds a git user prior to running run_build_github_release_package via github action

clpetersonucf committed:
Merge pull request #1324 from iturgeon/issue/release-action-fix-git-user

adds a git user prior to running run_build_github_release_package

iturgeon committed:
update heroku build process for new config options

iturgeon committed:
added missing AUTH_SALT to heroku app.json

iturgeon committed:
remove requirement to have THEME_PACKAGE set

iturgeon committed:
sets session and asset storage drivers for heroku

iturgeon committed:
attempt to fix storage driver config bug for heroku

clpetersonucf committed:
Merge pull request #1325 from iturgeon/issue/fix-heroku-after-config-update

fix heroku for config changes on v8.0.0

clpetersonucf committed:
github build-release can overwrite existing packages for a given tag

michaelferrara committed:
Adding to repo

michaelferrara committed:
Fixing github branch

michaelferrara committed:
Finished second pass through of my widgets page content

michaelferrara committed:
Made a few more general improvements to the MWP.

iturgeon commented on fuel/app/classes/controller/api/admin.php:
```diff
@@ -10,17 +10,21 @@ class Controller_Api_Admin extends Controller_Rest
 	public function get_widgets()
 	{
+		// must be super user for admin/widgets
+		if ( ! \Materia\Perm_Manager::is_super_user() ) throw new \HttpNotFoundException;
```
before() is called before every public controller function here.  [fuelphp before docs](https://fuelphp.com/docs/general/controllers/base.html#/before)

```suggestion
```

iturgeon commented on fuel/app/classes/controller/api/admin.php:
```diff
@@ -10,17 +10,21 @@ class Controller_Api_Admin extends Controller_Rest
 	public function post_widget($widget_id)
 	{
+		// must be super user for admin/widgets
+		if ( ! \Materia\Perm_Manager::is_super_user() ) throw new \HttpNotFoundException;
```
See above

```suggestion
```

iturgeon commented on fuel/app/classes/controller/api/admin.php:
```diff
@@ -64,4 +68,37 @@ public function post_user($user_id)
+
+	public function get_widget_search($input)
+	{
+		//no need to search if for some reason an empty string is passed
```
```suggestion
	{
		$input = trim($input);	
		//no need to search if for some reason an empty string is passed
```

iturgeon commented on fuel/app/classes/controller/api/admin.php:
```diff
@@ -64,4 +68,37 @@ public function post_user($user_id)
 		return \Service_User::update_user($user_id, $user);
 	}
+
+	public function get_widget_search($input)
```
need to test this when $input isn't set, but it'd be nice to add types
```suggestion
	public function get_widget_search(string $input)
```

iturgeon commented on fuel/app/classes/controller/api/admin.php:
```diff
@@ -64,4 +68,37 @@ public function post_user($user_id)
+		return \Materia\Widget_Instance_Manager::get_search($input);
+	}
+
+	public function get_extra_attempts($inst_id)
```
```suggestion
	public function get_extra_attempts(string $inst_id)
```

iturgeon commented on fuel/app/classes/controller/api/admin.php:
```diff
@@ -64,4 +68,37 @@ public function post_user($user_id)
+		return $inst->get_all_extra_attempts();
+	}
+
+	public function post_extra_attempts($inst_id)
```
```suggestion
	public function post_extra_attempts(string $inst_id)
```

iturgeon commented on fuel/app/classes/controller/api/admin.php:
```diff
@@ -10,17 +10,21 @@ class Controller_Api_Admin extends Controller_Rest
 
 	public function before()
```
This class should probably use `Trait_RateLimit`

iturgeon commented on fuel/app/classes/controller/api/admin.php:
```diff
@@ -64,4 +68,37 @@ public function post_user($user_id)
+
+	public function post_extra_attempts($inst_id)
+	{
+		// Validate input
```
```suggestion
		// Get POSTed json input
```

iturgeon commented on fuel/app/classes/controller/api/admin.php:
```diff
@@ -64,4 +68,37 @@ public function post_user($user_id)
+				$value['user_id'], 
+				$value['extra_attempts'], 
+				$value['context_id'],
+				$value['id'] > 0 ? $value['id'] : null);
```
extra_attempts values need better validation.

iturgeon commented on fuel/app/classes/controller/api/admin.php:
```diff
@@ -64,4 +68,37 @@ public function post_user($user_id)
+				$value['id'] > 0 ? $value['id'] : null);
+		}
+
+		return true;
```
this could return the updated values instead

iturgeon commented on fuel/app/classes/materia/api/v1.php:
```diff
@@ -42,7 +42,7 @@ static public function widgets_get_by_type($type)
 	}
 
-	static public function widget_instances_get($inst_ids = null)
+	static public function widget_instances_get($inst_ids = null, $deleted = false)
```
```suggestion
	static public function widget_instances_get($inst_ids = null, bool $deleted = false)
```

iturgeon commented on fuel/app/classes/materia/api/v1.php:
```diff
@@ -68,6 +68,17 @@ static public function widget_instance_delete($inst_id)
+	/**
+	 * @return bool, true if successfully undeleted widget instance, false otherwise.
+	 */
+	static public function widget_instance_undelete($inst_id)
```
```suggestion
	static public function widget_instance_undelete(string $inst_id)
```

iturgeon commented on fuel/app/classes/materia/api/v1.php:
```diff
@@ -68,6 +68,17 @@ static public function widget_instance_delete($inst_id)
+	static public function widget_instance_undelete($inst_id)
+	{
+		if ( ! Util_Validator::is_valid_hash($inst_id)) return Msg::invalid_input($inst_id);
+		if (\Service_User::verify_session() !== true) return Msg::no_login();
```
This isn't checking any permissions other then they are logged in - need to narrow this down significantly to:

* super user
* support user
* owners
  * what if one of the other owners deleted to prevent a co-owner from using? the co owner could just undelete?
  * what if a user accidentally deletes their own widget?
  * perhaps it's not worth letting users undelete?

iturgeon commented on fuel/app/classes/materia/api/v1.php:
```diff
@@ -68,6 +68,17 @@ static public function widget_instance_delete($inst_id)
+	{
+		if ( ! Util_Validator::is_valid_hash($inst_id)) return Msg::invalid_input($inst_id);
+		if (\Service_User::verify_session() !== true) return Msg::no_login();
+		if ( ! ($inst = Widget_Instance_Manager::get($inst_id, false, false, true))) return false; //TODO: THIS IS RETURNING FALSE
```
deal with TODO

iturgeon commented on fuel/app/classes/materia/perm/manager.php:
```diff
@@ -46,6 +46,25 @@ static public function is_super_user()
+	 * 
+	 * @return boolean wheter or not the current user has the support role as defined by Perm_Role class
+	 */
+	static public function is_support_user()
```
```suggestion
	static public function is_support_user(): bool
```

iturgeon commented on fuel/app/classes/materia/widget/instance.php:
```diff
@@ -14,6 +14,7 @@ class Widget_Instance
 	public $embed_url       = '';
 	public $is_student_made = false;
 	public $is_embedded     = false;
+	public $is_deleted			= false;
```
spaces between stuff, tabs to indent
```suggestion
	public $is_deleted      = false;
```

iturgeon commented on fuel/app/classes/materia/widget/instance.php:
```diff
@@ -387,6 +388,35 @@ public function db_remove()
+	 * 'Undeletes' instance by updating is_deleted flag from 1 to 0
+	 * @return bool true if successfully undeleted, false if unable to restore
+	 */
+	public function db_undelete()
```
add return type hint

```suggestion
	public function db_undelete(): bool
```

iturgeon commented on fuel/app/classes/materia/widget/instance.php:
```diff
@@ -387,6 +388,35 @@ public function db_remove()
+			->where('id', $this->id)
+			->execute();
+
+			// store an activity log
```
Too far indented

```suggestion
		// store an activity log
```

iturgeon commented on fuel/app/classes/materia/widget/instance.php:
```diff
@@ -550,6 +580,63 @@ public function lti_associations()
 			->get();
 	}
 
+	public function get_all_extra_attempts()
```
add return type hint

```suggestion
	public function get_all_extra_attempts(): array
```

iturgeon commented on fuel/app/classes/materia/widget/instance.php:
```diff
@@ -550,6 +580,63 @@ public function lti_associations()
+						])
+					->execute();
+			}
+		}
```
return something useful about the new state

iturgeon commented on fuel/app/classes/materia/widget/instance.php:
```diff
@@ -550,6 +580,63 @@ public function lti_associations()
+		return $result;
+	}
+
+	public function set_extra_attempts($user_id, $extra_attempts, $context_id, $id=null)
```
Consider a more specific id OR change to the unique key combo of user_id + inst_id 

```suggestion
	public function set_extra_attempts(int $user_id, int $extra_attempts, string $context_id, $id=null)
```

iturgeon commented on fuel/app/classes/materia/widget/instance/manager.php:
```diff
@@ -6,13 +6,13 @@ class Widget_Instance_Manager
 	public $validate = true;
 
-	static public function get($inst_id, $load_qset=false, $timestamp=false)
+	static public function get($inst_id, $load_qset=false, $timestamp=false, $deleted=false)
```
```suggestion
	static public function get(string $inst_id, bool $load_qset=false, $timestamp=false, bool $deleted=false): bool
```

iturgeon commented on fuel/app/classes/materia/widget/instance/manager.php:
```diff
@@ -6,13 +6,13 @@ class Widget_Instance_Manager
 	}
 
-	static public function get_all(Array $inst_ids, $load_qset=false, $timestamp=false)
+	static public function get_all(Array $inst_ids, $load_qset=false, $timestamp=false, $deleted=false)
```
```suggestion
	static public function get_all(Array $inst_ids, $load_qset=false, $timestamp=false, bool $deleted=false): array
```

iturgeon commented on fuel/app/classes/materia/widget/instance/manager.php:
```diff
@@ -45,6 +45,7 @@ static public function get_all(Array $inst_ids, $load_qset=false, $timestamp=fal
 				'open_at'         => $r['open_at'],
 				'close_at'        => $r['close_at'],
 				'attempts'        => $r['attempts'],
+				'is_deleted'			=> (bool) $r['is_deleted'],
```
Spaces for alignment, tabs for indenting.

iturgeon commented on fuel/app/classes/materia/widget/instance/manager.php:
```diff
@@ -101,4 +102,49 @@ public static function lock($inst_id)
+	 * 
+	 * @return array of widget instances related to the given input
+	 */
+	public static function get_search($input)
```
```suggestion
	public static function get_search(string $input): array
```

iturgeon commented on fuel/app/classes/materia/widget/instance/manager.php:
```diff
@@ -101,4 +102,49 @@ public static function lock($inst_id)
+		foreach ($results as $r)
+		{
+			$widget = new Widget();
+			$widget->get($r['widget_id']);
+			$inst = new Widget_Instance([
+				'id'              => $r['id'],
+				'user_id'         => $r['user_id'],
+				'name'            => $r['name'],
+				'is_student_made' => (bool) $r['is_student_made'],
+				'student_access'  => Perm_Manager::accessible_by_students($r['id'], Perm::INSTANCE),
+				'guest_access'    => (bool) $r['guest_access'],
+				'is_draft'        => (bool) $r['is_draft'],
+				'created_at'      => $r['created_at'],
+				'open_at'         => $r['open_at'],
+				'close_at'        => $r['close_at'],
+				'attempts'        => $r['attempts'],
+				'is_deleted'			=> (bool) $r['is_deleted'],
+				'embedded_only'   => (bool) $r['embedded_only'],
+				'widget'          => $widget,
+			]);
```
Move Perm_Manager::accessible_by_students before creation to clean up.
fix alignments

```suggestion
			$widget->get($r['widget_id']);
			$student_access = Perm_Manager::accessible_by_students($r['id'], Perm::INSTANCE);
			$inst = new Widget_Instance([
				'id'              => $r['id'],
				'user_id'         => $r['user_id'],
				'name'            => $r['name'],
				'is_student_made' => (bool) $r['is_student_made'],
				'student_access'  => $student_access,
				'guest_access'    => (bool) $r['guest_access'],
				'is_draft'        => (bool) $r['is_draft'],
				'created_at'      => $r['created_at'],
				'open_at'         => $r['open_at'],
				'close_at'        => $r['close_at'],
				'attempts'        => $r['attempts'],
				'is_deleted'      => (bool) $r['is_deleted'],
				'embedded_only'   => (bool) $r['embedded_only'],
				'widget'          => $widget,
			]);
```

iturgeon commented on fuel/app/themes/default/partials/header.php:
```diff
@@ -44,13 +44,15 @@
-						</li>
+						<?php if (\Materia\Perm_Manager::is_super_user()): ?>
+							<li>
+								<a class='elevated' href="/admin/widget">Widgets</a>
```
We probably need to have 2 menu items, one for support and one for admin.

iturgeon commented on src/components/extra-attempts-dialog.jsx:
```diff
@@ -0,0 +1,289 @@
+	)
+}
+
+export default ExtraAttemptsDialog
```
new line at end of file (check editor settings)
```suggestion
export default ExtraAttemptsDialog

```

iturgeon commented on src/components/extra-attempts-dialog.jsx:
```diff
@@ -0,0 +1,289 @@
+})
+
+// Component for each individual row in the Extra Attempts Gui
+const ExtraAttemptsRow = ({extraAttempt, user, onChange}) => {
```
Split off into separate component

iturgeon commented on src/components/support-selected-instance.jsx:
```diff
@@ -0,0 +1,353 @@
+	return s == 'true'
+}
+
+const PERM_VISIBLE = 1
```
Let's move this to a reusable constants file

iturgeon commented on src/components/support-search.jsx:
```diff
@@ -0,0 +1,109 @@
+	)
+}
+
+export default SupportSearch
```
newline at end of file (there's a bunch)

```suggestion
export default SupportSearch

```

iturgeon reviewed (changes requested):
I only had a chance to parse through some of the changes.  Most of my comments are on the server side (starting at top down).  So I started at the bottom up, but I think most of those are the support pages.   

Lots to go over, so I'll return later for more reviewing.

iturgeon commented on src/angular/controllers/ctrl-my-widgets-settings.js:
```diff
@@ -204,9 +206,9 @@ app.controller('MyWidgetsSettingsController', function (
+			; ({ date } = range)
 			const period = range.period ? range.period : 'am'
-			;({ time } = range)
+				; ({ time } = range)
```
These are super strange and hideous, would prefer a different syntax if prettier does this...

iturgeon commented on src/angular/services/srv-widget.js:
```diff
@@ -106,6 +106,10 @@ app.service('WidgetSrv', function (SelectedWidgetSrv, DateTimeServ, $q, $rootSco
+
+			console.log(args)
+			console.log(args)
+
```
```suggestion

```

iturgeon commented on src/angular/controllers/ctrl-my-widgets-settings.js:
```diff
@@ -136,6 +136,8 @@ app.controller('MyWidgetsSettingsController', function (
 				$scope.availability[i].anytime = true
 			}
+
+			console.log($scope.availability[i])
```
```suggestion
```

iturgeon commented on src/angular/controllers/ctrl-my-widgets-export.js:
```diff
@@ -6,6 +6,9 @@ app.controller('MyWidgetsExportController', function (Please, $scope, SelectedWi
 
+		console.log("==================================")
+		console.log($scope.selected)
+
```
```suggestion
```

iturgeon commented on src/components/extra-attempts-dialog.jsx:
```diff
@@ -0,0 +1,289 @@
+const fetchUsers = (arrayOfUserIds) => fetch('/api/json/user_get', fetchOptions({body: `data=${encodeURIComponent(JSON.stringify([arrayOfUserIds]))}`}))
+const searchUsers = (input) => fetch('/api/json/users_search', fetchOptions({body: `data=${encodeURIComponent(JSON.stringify([input]))}`}))
+const getExtraAttemptsForInstance = (instId) => fetch(`/api/admin/extra_attempts/${instId}`)
+const setExtraAttemptsForInstance = (instId, attempts) => fetch(`/api/admin/extra_attempts/${instId}`, 
```
suggest we centralize all of these api calls so we can abstract them & add consistent error handling.

This is something we could borrow from Obojobo Next or perhaps we should use react-query?

iturgeon commented on src/components/extra-attempts-dialog.jsx:
```diff
@@ -0,0 +1,289 @@
+	// sets row to display:none if removed, until parent render updates
+	const [disabled, setDisabled] = useState(false)
+
+	// anytime the state of this row changes, update the parent object
```
Kind of inventive use of useEffect to funnel changes into onChange.  I do wonder if it would be fine just calling onChange and updating state?

iturgeon commented on src/components/extra-attempts-dialog.jsx:
```diff
@@ -0,0 +1,289 @@
+				_users.forEach(u => { keyedUsers[u.id] = u })
+				setUsers(keyedUsers)
+			})
+		}, [inst]
```
More inspiration to extract/abstract queries.  would like to simplify this:

1. evaluate if we can use a single state update upon completion
2. handle errors
3. add loading indicators

iturgeon commented on src/components/extra-attempts-dialog.jsx:
```diff
@@ -0,0 +1,289 @@
+			<div className="extraAttemptsModal">
+				<span className="title">Give Students Extra Attempts</span>
+				<div className="attempts-container">
+					<div className="search-container">
```
Let's consolidate this user search interfaces into a single component for all the other uses

iturgeon commented on src/components/modal.jsx:
```diff
@@ -24,13 +21,15 @@ class Modal extends React.Component {
 		modalRoot.appendChild( this.element );
 		document.addEventListener('mousedown', this.clickOutsideListener);
-  		document.addEventListener('touchstart', this.clickOutsideListener);
+  	document.addEventListener('touchstart', this.clickOutsideListener);
```
mix of spaces and tabs 
```suggestion
		document.addEventListener('touchstart', this.clickOutsideListener);
```

iturgeon commented on src/components/my-widgets-copy-dialog.jsx:
```diff
@@ -1,27 +1,30 @@
 	const [copyPermissions, setCopyPermissions] = useState(false)
+
+	// Sets the initial title
+	useEffect(() => {setNewTitle(name + " copy")}, [])
```
Can I suggest using the parenthesis?

```suggestion
	useEffect(() => {setNewTitle(name + " (Copy)")}, [])
```

iturgeon commented on src/components/my-widgets-export.jsx:
```diff
@@ -0,0 +1,202 @@
+		if (checkAll.noChange === true) return
+
+		if (semesterOptions.length > 0) {
+			let arr = new Array(scores.length).fill(checkAll.check)
```
This is cool, but use `const` instead of let

iturgeon commented on src/components/my-widgets-export.jsx:
```diff
@@ -0,0 +1,202 @@
+	const [exportType, setExportType] = useState("Questions and Answers")
+	const [showOptions, setShowOptions] = useState(false)
+	const [semesterOptions, setSemesterOptions] = useState([])
+	const [checkAll, setCheckAll] = useState({check: false, noChange: false})
```
There may be a need to combine a handful of these into one useState variable.   I usually write like this on my first pass just to keep things straight, but there's some inconvenient inefficiencies that come from each value having it's own setState call.

Mainly, the problem is you must call multiple setStates in reaction to a single change in the component, causing multiple updates to that component.

It may be time to combine several state vars into a single object or use https://reactjs.org/docs/hooks-reference.html#usereducer

iturgeon commented on src/components/my-widgets-export.jsx:
```diff
@@ -0,0 +1,202 @@
+	}, [checkAll])
+
+	// Sets selected semesters and their respective header text
+	useEffect(() => {
```
I get the feeling we're using useEffect too much.  It's generally used for first time setup of a component OR for something that has a side effect on the DOM (direct, non-react DOM manipulation).

I think in this case useCallback may be more appropriate for a few of these....

iturgeon commented on src/components/my-widgets-export.jsx:
```diff
@@ -0,0 +1,202 @@
+	useEffect(() => {
+		if (semesterOptions.length > 0) {
+			let str = ""
+			let str_cpy = ""
```
```suggestion
			let str = "" // creates: 2020 Fall, 2020 Spring
			let str_cpy = "" // creates: 2020-Fall,2020-Spring
```

iturgeon commented on src/components/my-widgets-export.jsx:
```diff
@@ -0,0 +1,202 @@
+			}
+
+			setHeader(str)
+			setSelectedSemesters(str_cpy)
```
an example where it'd be best to call a single setState function

iturgeon commented on src/components/my-widgets-export.jsx:
```diff
@@ -0,0 +1,202 @@
+								{ 
+									exportOptions.map((val, index) => {
+										return (<option key={index} value={val}>{val}</option>)
+									})
```
take advantage of implicit arrow function object returns

```suggestion
									exportOptions.map((val, index) => <option key={index} value={val}>{val}</option>)
```

iturgeon commented on src/components/my-widgets-page.jsx:
```diff
@@ -48,21 +47,66 @@ const MyWidgetsPage = () => {
+		.catch(error => {
+			setIsLoading(false)
+			setWidgets([])
+		})
```
Combining of these setState calls is needed via a combined state or reducer.

iturgeon commented on src/components/my-widgets-score-semester-storage.jsx:
```diff
@@ -1,66 +1,212 @@
+const fetchStorageData = (instId) => fetch('/api/json/play_storage_get', fetchOptions({body:`data=%5B%22${instId}%22%5D`}))
 
 const defaultState = {
+	pageNumber: 0,
```
sheer size wise, we may want to break this component up, potentially pull the table off into a separate component?

iturgeon commented on src/components/my-widgets-score-semester-storage.jsx:
```diff
@@ -1,66 +1,212 @@
+	useEffect(() => {
+		if (state.isLoading === false)
+		{
+			let tmpVals = state.selectedTable.map((row, index) =>
```
prefer const everywhere it's able to be used

iturgeon commented on src/components/my-widgets-score-semester-storage.jsx:
```diff
@@ -1,66 +1,212 @@
+				${tmpVals.length} entries (filtered from ${Math.min(state.storageData?.length, MAX_ROWS)} total entries)`
+
+			setRowVals(tmpVals)
+			setState({...state, footerText: text})
```
avoid multiple setStates being called at once

iturgeon commented on src/components/my-widgets-score-semester-storage.jsx:
```diff
@@ -1,66 +1,212 @@
+	}, [semester, instId, searchInput])
+
+	// Sets the paginated rows and footer text
+	useEffect(() => {
```
this useEffect makes a pretty strong argument for moving the table into a sub component.

We'll just send the data as props and it can just render.

iturgeon commented on src/components/my-widgets-score-semester-storage.jsx:
```diff
@@ -1,66 +1,212 @@
+		searchInput
+	])
+
+	const onChangePageCount = useCallback(newValue => {
```
After some thinking about [over-optimizing hooks](https://kentcdodds.com/blog/usememo-and-usecallback), these may not be all that useful.  We're not passing them to child components and by having them change on state changes, we probably recreate them all the time anyway.

iturgeon commented on src/components/my-widgets-score-semester-storage.jsx:
```diff
@@ -1,66 +1,212 @@
+				"right"
+
+			switch(paginateType) {
+				case ("right"):
```
prefer no parenthesis 
```suggestion
				case "right":
```

iturgeon commented on src/components/my-widgets-score-semester-storage.jsx:
```diff
@@ -1,66 +1,212 @@
+		const sanatizedSearch = searchInput.replace(/\s+/g, '').toUpperCase()
+
+		if (searchInput.length === 0)
+			return true
```
avoid putting an `if` w/o curly brackets on 2 lines like the plague. Either a single line or use curly brackets.

Reason: it's just too easy for future devs to do something unintentionally here
```suggestion
		if (searchInput.length === 0) return true
```

iturgeon commented on src/components/my-widgets-score-semester-storage.jsx:
```diff
@@ -1,66 +1,212 @@
+			return true
+
+		// Matches by user
+		if (val.play.user.replace(/\s+/g, '').toUpperCase().indexOf(sanatizedSearch) > -1)
```
```suggestion
		if (val.play.user.replace(/\s+/g, '').toUpperCase().includes(sanatizedSearch))
```

iturgeon commented on src/components/my-widgets-score-semester-storage.jsx:
```diff
@@ -1,66 +1,212 @@
+			return true
+
+		// Matches by first and last
+		if (firstLast.replace(/\s+/g, '').toUpperCase().indexOf(sanatizedSearch) > -1)
```
```suggestion
		if (firstLast.replace(/\s+/g, '').toUpperCase().includes(sanatizedSearch))
```

iturgeon commented on src/components/my-widgets-scores.jsx:
```diff
@@ -1,60 +1,84 @@
+	const closeExport = () => {
+		setState({...state, showExport: false})
+		document.body.style.overflow = "auto"
+	}
```
seeing a `document.something = something` set's off my "side effect" warning.  this should probably be a useEffect

```suggestion
	useEffect(() => {
		document.body.style.overflow = state.showExport ? 'hidden' : 'auto'
	}, [state.showExport]) 
	
	const openExport = () => {
		setState({...state, showExport: true})
	}

	const closeExport = () => {
		setState({...state, showExport: false})
	}
```

iturgeon commented on src/components/my-widgets-selected-instance.jsx:
```diff
@@ -1,328 +1,423 @@
-}
-
-export default MyWidgetSelectedInstance
+import React, { useEffect, useRef, useCallback, useState, useMemo} from 'react'
```
somethings up with the diff on this file.  First guess is the line endings changed (perhaps edited on Windows?)

I'll skip till the diff is more usefull

iturgeon commented on src/components/my-widgets-settings-dialog.jsx:
```diff
@@ -0,0 +1,728 @@
+import React, { useState, useEffect } from 'react'
```
a monster, honkin' component.  this one's going to need some splitting up and optimization.  

1. move any functions we can outside the render loop (looks like a handful here that only operate on their arguments, not relying on scope)
2. `document.anything` is a concern, might need some refs?
3. some functions w/ multiple setStates, try to avoid
4. consider creating components for individual controls

iturgeon commented on src/components/my-widgets-settings-dialog.jsx:
```diff
@@ -0,0 +1,728 @@
+const fetchUpdateWidget = (args) => fetch('/api/json/widget_instance_update', fetchOptions({body: `data=${encodeURIComponent(JSON.stringify(args))}`}))
+const fetchUsers = (arrayOfUserIds) => fetch('/api/json/user_get', fetchOptions({body: `data=${encodeURIComponent(JSON.stringify([arrayOfUserIds]))}`}))
+
+const initForm = {
```
be cautious about using default/initial settings like this.
since this is pass by ref, if the component changes properties of these they may bleed to new instances of the component? 

One defense is to create a function that returns a brand new object instead

```javascript
const defaultForm = () => ({
  key: value
})

...

const [formData, setFormData] = useState(defaultForm())


```

iturgeon commented on src/components/support-search.jsx:
```diff
@@ -0,0 +1,109 @@
+	}, [debouncedSearchTerm])
+
+	const setSearchVal = (e) => {
+		if (!isSearching && e.target.value  !== "")
+			setIsSearching(true)
+
+		if (e.target.value === "")
+			setIsSearching(false)
```
say no to multi line, bracket-less `if` statements
```suggestion
		if (!isSearching && e.target.value  !== "") setIsSearching(true)
		if (e.target.value === "") setIsSearching(false)
```

iturgeon commented on src/components/support-search.jsx:
```diff
@@ -0,0 +1,109 @@
+				})
+				.then(instances => 
+				{
+					setSearchResults(instances)
+					setIsSearching(false)
```
seek a single setState call by combining state or using a reducer

iturgeon commented on src/components/support-search.jsx:
```diff
@@ -0,0 +1,109 @@
+			searchWidgets(debouncedSearchTerm)
+				.then(resp => {
+					// no content
+					if(resp.status === 502 || resp.status === 204) {
+						setIsSearching(false)
+						return []
+					}
```
duplicate comment, but we should consolidate a universal rest call interface & error handling.

iturgeon reviewed (changes requested):
Finished my pass over the PR @michaelferrara hopefully this second one is more pertinent to your efforts

michaelferrara committed:
Added Trait_RateLimit to admin.php

michaelferrara committed:
Removed Trait_RateLimit

Misread Ian's comment and thought it was on this file. Still getting used to PR system.

michaelferrara committed:
Added Trait_RateLimit

michaelferrara committed:
Adding newline to the end of file.

Adding newline to loading-icon.jsx

michaelferrara committed:
Applying multiple suggestions from Ian

Went through and tested/applied all the suggestions I could do quickly.

Co-authored-by: Ian Turgeon <iturgeon@hey.com>

michaelferrara committed:
Added newline to end of SupportSelectedInstance

michaelferrara committed:
Adding changes so I can merge with PR

michaelferrara committed:
Merge branch 'issue/support-dashboard-in-react' of https://github.com/michaelferrara/Materia into issue/support-dashboard-in-react

michaelferrara committed:
Added react query, and testing no content error.

iturgeon committed:
adds prodction ready nginx conf

iturgeon committed:
Merge remote-tracking branch 'upstream/master' into issue/production-ready-docker

iturgeon committed:
update composer, add ucf theme, dockerfile

iturgeon committed:
adds ucf dockerfile

iturgeon committed:
simplify ucf docker build by downloading auth module prior to docker build

iturgeon committed:
remove private ucf auth package

iturgeon committed:
adds docker image create action

- renames docker-compose images to -dev
- adds github action that builds and publishes docker images when a tag is pushed

iturgeon committed:
changes needed to get the tests to run

iturgeon committed:
test

iturgeon committed:
get commit hook working with docker

iturgeon committed:
consolidate gh actions

iturgeon committed:
get pre-commit to exit on failure

iturgeon committed:
minor syntax fixup

iturgeon committed:
xdebug removed, using dev env for dev

iturgeon committed:
updates to readme

iturgeon committed:
sets dev log threshold to allow debug messages to come through

iturgeon committed:
updates docker readme tabs/spaces

iturgeon committed:
Merge pull request #1334 from iturgeon/issue/production-ready-docker

Changes to make Docker in production possible.

iturgeon committed:
removes skip_checks gh action

iturgeon committed:
fixes run_build_github_release_package after docker changes

iturgeon committed:
update name of docker images and release zip job

iturgeon committed:
github release build change expanded zip file name

michaelferrara committed:
Added accessibility indicator, widget player, and various bug and content fixes

michaelferrara committed:
Adding tests and final comments and changes

FrenjaminBanklin committed:
Merged in and (hopefully) corrected upstream changes.

FrenjaminBanklin committed:
Renamed .jsx file to .js since it wasn't using React. Minor corrections to indentation/trailing whitespaces in a handful of files. Refactored parts of API utility to standardize construction of fetch body data strings. Fixed bug in perm-to-object collation function that was preventing the Edit Widget button from working when it should. Fixed widget player only working if the page was running from 127.0.0.1. Rolled back argument type expectations in widget instance manager (for now) to fix failing tests.

FrenjaminBanklin committed:
(Hopefully) Brought the remaining Angular frontend code up to date with divergent changes from the MSCA to restore functionality for the question importer and qset history tools in the widget creator.

FrenjaminBanklin committed:
In-progress: replacing jqueryDataTable plugin with a custom-built React data table component for the question importer.

FrenjaminBanklin committed:
Finished React question importer component.

FrenjaminBanklin committed:
Minor improvements to React question importer component and associated API call to prevent constant attempts when no questions are available from the API, and to indicate the various 'no questions' status messages.

FrenjaminBanklin committed:
Fixed the player/player page components to handle embedded previews. Cleaned up the homepage for code readability.

FrenjaminBanklin committed:
Merged dev branch, resolved conflicts.

FrenjaminBanklin committed:
Restoring development config so WDS assets work from localhost again - pending a better solution.

FrenjaminBanklin committed:
Cleaned up indentation or trailing whitespace in numerous files. Started removing unnecessary frontend assets. Fixed bug in My Widgets collaboration window where removing a user's permissions to a window then opening the collaboration window again then adding the same user would throw an error. Fixed bug in My Widgets page where using cached data would cause old widgets to be selected/editable until the fetch finished, and new widgets would not be selected automatically once the fetch finished. Refactored code to make conditional renders more readable.

FrenjaminBanklin committed:
Cleaned up trailing whitespace and indentation in numerous files. Replaced leftover hard-coded URI-encoded strings in React API utility. Re-implemented highlighting matching search strings in My Widgets page. Fixed individual score links in My Widgets page. Fixed low-resolution Materia logo being used in the React header component.

FrenjaminBanklin committed:
Removed beard mode service since all that is handled in React now. Fixed part of the test setup that was still expecting Angular filters, all of which have been deprecated. Re-engineered how accessibility options are parsed and displayed to be less error-prone.

FrenjaminBanklin committed:
Removed unnecessary snapshot and unused sass file. Corrected erroneous require in an Angular test. Fixed mobile-friendly filter list not working properly in the widget catalog. Fixed filters for keyboard/screen reader accessibility following changes to how those are processed elsewhere. Refactored various files for readability.

FrenjaminBanklin committed:
Fixing broken links to individual scores from user admin panel. Removing unused copies of Sass files.

FrenjaminBanklin committed:
React code cleanup.

FrenjaminBanklin committed:
Regular cleanup in numerous files, fixed setting expiration dates for collaborators.

FrenjaminBanklin committed:
More cleanup. Made the behavior after copying/deleting a widget in the My Widgets page a bit more stable. Fixed the 'Grant Access...' checkbox being ignored when making widget instance copies.

FrenjaminBanklin committed:
Mostly just cleanup this time.

FrenjaminBanklin committed:
More or less finished refactoring React code. Removed a few more deprecated files.

FrenjaminBanklin committed:
Removed hard-coded paths for images in templates/components/scss.
